components/raftstore/src/store/peer.rs

```diff
@@ -381,6 +392,13 @@ impl<S: Snapshot> CmdEpochChecker<S> {
             }
             self.proposed_admin_cmd
                 .push_back(ProposedAdminCmd::new(cmd_type, epoch_state, index));
+        } else if epoch_state.check_ver || epoch_state.check_conf_ver {
+            let proposed = ProposedAdminCmd::new(cmd_type, epoch_state, index);
+            // Although the command does not change the epoch, it should be checked whether
+            // it is mutually exclusive with other commands which change the epoch.
+            if proposed.is_mutually_exclusive() {
```

Comment:
Leaders do not always propose `AdminCmdType::TransferLeader`. And actual transfer leader is initiated after transferee applied this cmd which makes the check here too early to prevent issue #12410.

https://github.com/tikv/tikv/blob/aa5fb36af8e942bb136fc285a5498adf80be9c8c/components/raftstore/src/store/fsm/peer.rs#L3721-L3748

Reply:
How about supplying extra check in `ready_to_transfer_leader` to prevent this problem ?
https://github.com/tikv/tikv/blob/aa5fb36af8e942bb136fc285a5498adf80be9c8c/components/raftstore/src/store/peer.rs#L3963